flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TablePlanner.java

```diff
@@ -0,0 +1,139 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.maintenance.operator;
+
+import java.util.concurrent.ExecutorService;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.functions.OpenContext;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.streaming.api.functions.ProcessFunction;
+import org.apache.flink.util.Collector;
+import org.apache.iceberg.MetadataTableType;
+import org.apache.iceberg.MetadataTableUtils;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.flink.maintenance.api.DeleteOrphanFiles;
+import org.apache.iceberg.flink.maintenance.api.Trigger;
+import org.apache.iceberg.flink.source.FlinkSplitPlanner;
+import org.apache.iceberg.flink.source.ScanContext;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplitSerializer;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Plans the splits to read a metadata table content. */
+@Internal
+public class TablePlanner extends ProcessFunction<Trigger, TablePlanner.SplitInfo> {
+  private static final Logger LOG = LoggerFactory.getLogger(TablePlanner.class);
+
+  private final String taskName;
+  private final int taskIndex;
+  private final TableLoader tableLoader;
+  private final int workerPoolSize;
+  private final ScanContext scanContext;
+  private transient ExecutorService workerPool;
+  private transient Counter errorCounter;
+  private transient Table table;
+  private final boolean caseSensitive;
+  private transient IcebergSourceSplitSerializer splitSerializer;
+
+  public TablePlanner(
+      String taskName,
+      int taskIndex,
+      TableLoader tableLoader,
+      ScanContext scanContext,
+      int workerPoolSize,
+      boolean caseSensitive) {
+    Preconditions.checkNotNull(taskName, "Task name should no be null");
+    Preconditions.checkNotNull(tableLoader, "Table should no be null");
+    Preconditions.checkArgument(scanContext.isStreaming(), "Streaming should be set to true");
+
+    this.taskName = taskName;
+    this.taskIndex = taskIndex;
+    this.tableLoader = tableLoader;
+    this.scanContext = scanContext;
+    this.workerPoolSize = workerPoolSize;
+    this.caseSensitive = caseSensitive;
+  }
+
+  @Override
+  public void open(OpenContext openContext) throws Exception {
+    tableLoader.open();
+    Table tableOri = tableLoader.loadTable();
+    this.table =
+        MetadataTableUtils.createMetadataTableInstance(tableOri, MetadataTableType.ALL_FILES);
```

Comment:
Maybe extract this from the operator, and provide this in the constructor.
It would be nice if this planner could remain independent of which table we read, and the query/table could be provided as a parameter